feat: local image handling

brainbicycle opened the pull request:
The type of this PR is: **feat**

<!-- Bugfix/Feature/Enhancement/Documentation -->

<!-- If applicable, write the Jira ticket number in square brackets e.g. [PROJECT-XXXX]
     The Jira integration will turn it into a clickable link for you. -->

This PR resolves [CX-2014]

### Description

<!-- Implementation description -->

Adds ability to store and retrieve multiple local images for my collection works. 

1. Adds a [local image store](src/lib/utils/LocalImageStore.ts) that uses async storage and stores the path to a local image as well as height and width
2. Adds some [my collection specific image utils](src/lib/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionImageUtil.ts) that store my collection images with a key made up of the artwork id and the image index, allowing to store multiple per artwork
3. On adding a new my collection artwork stores image locally
4. On deleting images deletes local images
5. Adds some image views that use local images when remote images are not present, used in list and details view in My Collection

Follow-ups:
- should we be clearing the async storage periodically? this is small amount of data but maybe?
- edit case is still not really handled robustly
- although we are storing width and height the local image is using react-native image view rather than OpaqueImageView so they render pretty differently
- ~should combine the two image views they are pretty similar~

<details><summary>How it looks</summary>

https://user-images.githubusercontent.com/49686530/146552520-ca20f1eb-6117-46cf-bdc5-535f986b1cae.mp4

</details>

### PR Checklist (tick all before merging)

<!-- 💡 This checklist is experimental. CX warmly welcomes any feedback about the list or how it impacts your workflow -->

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [ ] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests/stories for my changes, or my changes don't require testing/stories, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/main/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/main/docs/adding_state_migrations.md))
- [x] I have added a changelog entry below or my changes do not require one.

### To the reviewers 👀

- [x] I would like at least one of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->

#### Cross-platform user-facing changes

- Allow storing and retrieving local images in my collection - Brian

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>


[CX-2014]: https://artsyproduct.atlassian.net/browse/CX-2014?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

brainbicycle commented on ArtsyWidget/Assets.xcassets/BlackArtsyLogo.imageset/Contents.json:
```diff
@@ -1,23 +1,23 @@
-      "filename" : "android-chrome-512x512.png",
-      "idiom" : "universal",
-      "scale" : "1x"
+      "filename": "android-chrome-512x512.png",
```
remove these formatting changes

ArtsyOpenSource commented:
<!--
  0 failure: 
  1 warning:  ❌ **An error occu...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  ❌ **An error occurred while validating your changelog, please make sure you provided a valid changelog.**
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2b4145de5d2c4a275be74757bf61c5c1fea8bb3a
</p>

brainbicycle commented on src/lib/Scenes/MyCollection/Components/MyCollectionImageView.tsx:
```diff
@@ -0,0 +1,58 @@
+
+  const [localImage, setLocalImage] = useState<LocalImage | null>(null)
+
+  useEffect(() => {
```
TODO: Should combine this and the MyCollectionDetailsImageView, they are very similar

lordkiz replied:
I thought same too. Do we have a followup ticket?

brainbicycle replied:
let me just do it this afternoon actually, shouldn't be too hard I was being lazy

brainbicycle replied:
alright I did this piece, I'll look into serializing an object today

brainbicycle commented on src/lib/Scenes/MyCollection/MyCollection.tsx:
```diff
@@ -329,6 +334,18 @@ const MyCollection: React.FC<{
+          <Button
+            onPress={async () => {
+              toast.show("Adding artwork", "middle")
+              await addRandomMyCollectionArtwork()
```
little dev utility to quickly add my collection items

lordkiz replied:
🚀

ArtsyOpenSource commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



### This PR contains the following changes:

- Cross-platform user-facing changes (Allow storing and retrieving local images in my collection - Brian)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e9601e7b46147cab8dc53ff8d0f323f2dea8bd68
</p>

lordkiz commented on src/lib/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -239,18 +197,59 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
+  refreshMyCollection()
+  props.onSuccess()
+  setTimeout(() => {
+    GlobalStore.actions.myCollection.artwork.resetForm()
```
Maybe we can call this before `props.onSuccess` and so avoid using a timeout for this? Would doing this lead to a situation where you'd be popping a page that is in the process of rerendering, since `resetForm()` would cause a rerender? I am not sure.

brainbicycle replied:
hmm, I think maybe this was old code and got removed in main, I don't see it anymore and I certainly didn't mean to add it. Removing here, good catch!

brainbicycle commented:
Alright updated to store objects rather than multiple keys which is cleaner, thanks for suggestion!

Also believe I fixed the multiple image issue, added another little dev utility to view local images for an artwork, here it is:
https://user-images.githubusercontent.com/49686530/147248599-b51c5599-a4b6-4201-904c-173e2557e38d.mp4

I want to take another look at that handle submit method you commented on just to make sure I didn't change anything there and then this should be good to go.

lordkiz reviewed (approved):
Looks great to me. 🚀
